```
Error running mint automation
+ mkdir -p 13976-f02a48a/gopath/src/github.com/minio
+ git -C 13976-f02a48a/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 13976-f02a48a/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 13976-f02a48a/gopath/src/github.com/minio/minio fetch --quiet minio pull/13976/head:pr13976
+ git -C 13976-f02a48a/gopath/src/github.com/minio/minio checkout --quiet pr13976
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/13976-f02a48a/gopath
+ make -C 13976-f02a48a/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.2
go: downloading github.com/colinmarc/hdfs/v2 v2.2.0
go: downloading github.com/minio/minio-go/v7 v7.0.17
go: downloading github.com/minio/cli v1.22.0
go: downloading github.com/minio/pkg v1.1.9
go: downloading google.golang.org/api v0.58.0
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading cloud.google.com/go/storage v1.10.0
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/Azure/azure-storage-blob-go v0.10.0
go: downloading golang.org/x/sys v0.0.0-20211020174200-9d6173849985
go: downloading github.com/Azure/azure-pipeline-go v0.2.2
go: downloading github.com/buger/jsonparser v1.1.1
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/rs/dnscache v0.0.0-20210201191234-295bba877686
go: downloading golang.org/x/crypto v0.0.0-20211209193657-4570a0811e8b
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading github.com/fatih/color v1.13.0
go: downloading github.com/golang-jwt/jwt/v4 v4.1.0
go: downloading github.com/minio/kes v0.14.0
go: downloading github.com/klauspost/compress v1.13.6
go: downloading github.com/Shopify/sarama v1.27.2
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading github.com/minio/sio v0.3.0
go: downloading github.com/VividCortex/ewma v1.1.1
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/cosnicolaou/pbzip2 v1.0.1
go: downloading github.com/miekg/dns v1.1.43
go: downloading github.com/minio/console v0.12.9
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/dchest/siphash v1.2.1
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/inconshreveable/mousetrap v1.0.0
go: downloading github.com/bits-and-blooms/bloom/v3 v3.0.1
go: downloading github.com/go-openapi/loads v0.20.2
go: downloading github.com/klauspost/reedsolomon v1.9.15
go: downloading github.com/klauspost/readahead v1.3.1
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/minio/selfupdate v0.4.0
go: downloading github.com/minio/zipindex v0.2.1
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/philhofer/fwd v1.1.2-0.20210722190033-5c56ac6d0bb9
go: downloading github.com/pierrec/lz4 v2.6.0+incompatible
go: downloading github.com/prometheus/client_golang v1.11.0
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading github.com/prometheus/procfs v0.7.3
go: downloading github.com/rs/cors v1.7.0
go: downloading github.com/shirou/gopsutil/v3 v3.21.9
go: downloading github.com/tinylib/msgp v1.1.7-0.20211026165309-e818a1881b0e
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/yargevad/filepathx v1.0.0
go: downloading github.com/zeebo/xxh3 v1.0.0
go: downloading go.etcd.io/etcd/api/v3 v3.5.0
go: downloading go.etcd.io/etcd/client/v3 v3.5.0
go: downloading go.uber.org/atomic v1.9.0
go: downloading github.com/go-ldap/ldap/v3 v3.2.4
go: downloading go.uber.org/zap v1.19.1
go: downloading github.com/coredns/coredns v1.4.0
go: downloading github.com/beevik/ntp v0.3.0
go: downloading github.com/golang-jwt/jwt v3.2.2+incompatible
go: downloading github.com/minio/simdjson-go v0.2.1
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/eclipse/paho.mqtt.golang v1.3.0
go: downloading github.com/elastic/go-elasticsearch/v7 v7.12.0
go: downloading github.com/go-sql-driver/mysql v1.5.0
go: downloading github.com/gomodule/redigo v1.8.5
go: downloading github.com/lib/pq v1.9.0
go: downloading github.com/nats-io/nats.go v1.11.1-0.20210623165838-4b75fc59ae30
go: downloading github.com/nats-io/stan.go v0.8.3
go: downloading github.com/nsqio/go-nsq v1.0.8
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/streadway/amqp v1.0.0
go: downloading golang.org/x/time v0.0.0-20210723032227-1f47c861a9ac
go: downloading github.com/golang/protobuf v1.5.2
go: downloading golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading github.com/mattn/go-colorable v0.1.10
go: downloading github.com/mattn/go-isatty v0.0.14
go: downloading github.com/xdg/stringprep v1.0.0
go: downloading github.com/mattn/go-runewidth v0.0.13
go: downloading github.com/minio/csvparser v1.0.0
go: downloading github.com/minio/parquet-go v1.1.0
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading github.com/klauspost/cpuid/v2 v2.0.9
go: downloading golang.org/x/net v0.0.0-20211112202133-69e39bad7dc2
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.0
go: downloading google.golang.org/grpc v1.41.0
go: downloading cloud.google.com/go v0.94.1
go: downloading github.com/gorilla/websocket v1.4.2
go: downloading github.com/bits-and-blooms/bitset v1.2.0
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/prometheus/common v0.31.1
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/beorn7/perks v1.0.1
go: downloading github.com/golang/snappy v0.0.3
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20200615164410-66371956d46c
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.1
go: downloading github.com/go-openapi/analysis v0.20.0
go: downloading github.com/go-openapi/spec v0.20.3
go: downloading github.com/go-openapi/swag v0.19.14
go: downloading google.golang.org/genproto v0.0.0-20210928142010-c7af6a1a74c9
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/jcmturner/gofork v1.0.0
go: downloading github.com/hashicorp/go-uuid v1.0.2
go: downloading go.uber.org/multierr v1.7.0
go: downloading golang.org/x/text v0.3.7
go: downloading github.com/rivo/uniseg v0.2.0
go: downloading github.com/nats-io/nkeys v0.3.0
go: downloading github.com/rs/xid v1.3.0
go: downloading google.golang.org/protobuf v1.27.1
go: downloading gopkg.in/ini.v1 v1.63.2
go: downloading github.com/minio/md5-simd v1.1.2
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/eapache/go-resiliency v1.2.0
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/rcrowley/go-metrics v0.0.0-20200313005456-10cdbea86bc0
go: downloading gopkg.in/jcmturner/gokrb5.v7 v7.5.0
go: downloading github.com/lestrrat-go/jwx v1.2.7
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/fatih/structs v1.1.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading github.com/coreos/go-systemd/v22 v22.3.2
go: downloading github.com/mailru/easyjson v0.7.6
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2-0.20181231171920-c182affec369
go: downloading github.com/jcmturner/goidentity/v6 v6.0.1
go: downloading github.com/apache/thrift v0.15.0
go: downloading github.com/tidwall/sjson v1.2.3
go: downloading github.com/tidwall/gjson v1.11.0
go: downloading github.com/josharian/intern v1.0.0
go: downloading github.com/tidwall/pretty v1.2.0
go: downloading github.com/tidwall/match v1.1.1
go: downloading github.com/go-openapi/strfmt v0.20.0
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/lestrrat-go/iter v1.0.1
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.8
go: downloading github.com/lestrrat-go/blackmagic v1.0.0
go: downloading github.com/lestrrat-go/httpcc v1.0.0
go: downloading github.com/go-openapi/jsonreference v0.19.5
go: downloading github.com/asaskevich/govalidator v0.0.0-20200907205600-7a23bdc65eef
go: downloading go.mongodb.org/mongo-driver v1.4.6
go: downloading github.com/mitchellh/mapstructure v1.4.1
go: downloading github.com/go-openapi/errors v0.19.9
go: downloading github.com/PuerkitoBio/purell v1.1.1
go: downloading github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578
go: downloading github.com/googleapis/gax-go/v2 v2.1.1
go: downloading go.opencensus.io v0.23.0
go: downloading github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e
go: downloading github.com/go-stack/stack v1.8.0
go: downloading github.com/go-openapi/runtime v0.19.24
go: downloading github.com/jessevdk/go-flags v1.4.0
go: downloading github.com/minio/operator v0.0.0-20211011212245-31460bbbc4b7
go: downloading github.com/minio/mc v0.0.0-20211207230606-23a05f5a17f2
go: downloading github.com/minio/operator/logsearchapi v0.0.0-20211011212245-31460bbbc4b7
go: downloading github.com/unrolled/secure v1.0.9
go: downloading k8s.io/client-go v0.21.1
go: downloading golang.org/x/oauth2 v0.0.0-20210819190943-2bc19b11175f
go: downloading github.com/minio/direct-csi v1.3.5-0.20210601185811-f7776f7961bf
go: downloading github.com/go-openapi/validate v0.20.2
go: downloading github.com/docker/go-units v0.4.0
go: downloading k8s.io/apimachinery v0.21.1
go: downloading k8s.io/kube-openapi v0.0.0-20210305001622-591a79e4bda7
go: downloading github.com/mb0/glob v0.0.0-20160210091149-1eb79d2de6c4
go: downloading github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
go: downloading gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: downloading gopkg.in/jcmturner/aescts.v1 v1.0.1
go: downloading gopkg.in/jcmturner/rpc.v1 v1.1.0
go: downloading maze.io/x/duration v0.0.0-20160924141736-faac084b6075
go: downloading github.com/minio/colorjson v1.0.1
go: downloading github.com/posener/complete v1.2.3
go: downloading gopkg.in/h2non/filetype.v1 v1.0.5
go: downloading golang.org/x/term v0.0.0-20210927222741-03fcf44c2211
go: downloading github.com/minio/filepath v1.0.0
go: downloading github.com/pkg/profile v1.6.0
go: downloading github.com/briandowns/spinner v1.16.0
go: downloading github.com/tklauser/go-sysconf v0.3.9
go: downloading github.com/georgysavva/scany v0.2.7
go: downloading k8s.io/api v0.21.1
go: downloading k8s.io/klog/v2 v2.8.0
go: downloading github.com/pkg/xattr v0.4.3
go: downloading sigs.k8s.io/controller-runtime v0.8.0
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/go-logr/logr v0.4.0
go: downloading github.com/tklauser/numcpus v0.3.0
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/stretchr/testify v1.7.0
go: downloading sigs.k8s.io/yaml v1.2.0
go: downloading sigs.k8s.io/structured-merge-diff/v4 v4.1.0
go: downloading github.com/google/gofuzz v1.1.0
go: downloading github.com/google/go-cmp v0.5.6
go: downloading gopkg.in/inf.v0 v0.9.1
go: downloading github.com/stretchr/objx v0.2.0
go: downloading github.com/pmezard/go-difflib v1.0.1-0.20181226105442-5d4384ee4fb2
go: downloading gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
go: downloading github.com/emicklei/go-restful v2.9.5+incompatible
go: downloading k8s.io/utils v0.0.0-20201110183641-67b214c5f920
go: downloading github.com/googleapis/gnostic v0.5.1
../../../../pkg/mod/github.com/minio/operator@v0.0.0-20211011212245-31460bbbc4b7/pkg/apis/minio.min.io/v1/helper.go:49:2: github.com/minio/madmin-go@v1.1.18: replacement directory ../madmin-go does not exist
make: *** [build] Error 1

```

--- a/cmd/admin-handlers-users.go
+++ b/cmd/admin-handlers-users.go
@@ -416,14 +416,14 @@ func (a adminAPIHandlers) AddUser(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	var uinfo madmin.UserInfo
-	if err = json.Unmarshal(configBytes, &uinfo); err != nil {
+	var ureq madmin.AddOrUpdateUserReq
+	if err = json.Unmarshal(configBytes, &ureq); err != nil {
 		logger.LogIf(ctx, err)
 		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrAdminConfigBadJSON), r.URL)
 		return
 	}
 
-	if err = globalIAMSys.CreateUser(ctx, accessKey, uinfo); err != nil {
+	if err = globalIAMSys.CreateUser(ctx, accessKey, ureq); err != nil {
 		writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
 		return
 	}

--- a/cmd/admin-handlers-users_test.go
+++ b/cmd/admin-handlers-users_test.go
@@ -18,8 +18,15 @@
 package cmd
 
 import (
+	"bytes"
 	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
 	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
 	"os"
 	"strings"
 	"testing"
@@ -29,7 +36,9 @@ import (
 	minio "github.com/minio/minio-go/v7"
 	"github.com/minio/minio-go/v7/pkg/credentials"
 	cr "github.com/minio/minio-go/v7/pkg/credentials"
+	"github.com/minio/minio-go/v7/pkg/s3utils"
 	"github.com/minio/minio-go/v7/pkg/set"
+	"github.com/minio/minio-go/v7/pkg/signer"
 	"github.com/minio/minio/internal/auth"
 )
 
@@ -177,6 +186,7 @@ func TestIAMInternalIDPServerSuite(t *testing.T) {
 
 				suite.SetUpSuite(c)
 				suite.TestUserCreate(c)
+				suite.TestUserPolicyEscalationBug(c)
 				suite.TestPolicyCreate(c)
 				suite.TestCannedPolicies(c)
 				suite.TestGroupAddRemove(c)
@@ -222,6 +232,24 @@ func (s *TestSuiteIAM) TestUserCreate(c *check) {
 		c.Fatalf("user could not create bucket: %v", err)
 	}
 
+	// 3.10. Check that user's password can be updated.
+	_, newSecretKey := mustGenerateCredentials(c)
+	err = s.adm.SetUser(ctx, accessKey, newSecretKey, madmin.AccountEnabled)
+	if err != nil {
+		c.Fatalf("Unable to update user's secret key: %v", err)
+	}
+	// 3.10.1 Check that old password no longer works.
+	err = client.MakeBucket(ctx, getRandomBucketName(), minio.MakeBucketOptions{})
+	if err == nil {
+		c.Fatalf("user was unexpectedly able to create bucket with bad password!")
+	}
+	// 3.10.2 Check that new password works.
+	client = s.getUserClient(c, accessKey, newSecretKey, "")
+	err = client.MakeBucket(ctx, getRandomBucketName(), minio.MakeBucketOptions{})
+	if err != nil {
+		c.Fatalf("user could not create bucket: %v", err)
+	}
+
 	// 4. Check that user can be disabled and verify it.
 	err = s.adm.SetUserStatus(ctx, accessKey, madmin.AccountDisabled)
 	if err != nil {
@@ -257,6 +285,108 @@ func (s *TestSuiteIAM) TestUserCreate(c *check) {
 	err = client.MakeBucket(ctx, getRandomBucketName(), minio.MakeBucketOptions{})
 	if err == nil {
 		c.Fatalf("user account was not deleted!")
+	}
+}
+
+func (s *TestSuiteIAM) TestUserPolicyEscalationBug(c *check) {
+	ctx, cancel := context.WithTimeout(context.Background(), testDefaultTimeout)
+	defer cancel()
+
+	bucket := getRandomBucketName()
+	err := s.client.MakeBucket(ctx, bucket, minio.MakeBucketOptions{})
+	if err != nil {
+		c.Fatalf("bucket creat error: %v", err)
+	}
+
+	// 2. Create a user, associate policy and verify access
+	accessKey, secretKey := mustGenerateCredentials(c)
+	err = s.adm.SetUser(ctx, accessKey, secretKey, madmin.AccountEnabled)
+	if err != nil {
+		c.Fatalf("Unable to set user: %v", err)
+	}
+	// 2.1 check that user does not have any access to the bucket
+	uClient := s.getUserClient(c, accessKey, secretKey, "")
+	c.mustNotListObjects(ctx, uClient, bucket)
+
+	// 2.2 create and associate policy to user
+	policy := "mypolicy-test-user-update"
+	policyBytes := []byte(fmt.Sprintf(`{
+ "Version": "2012-10-17",
+ "Statement": [
+  {
+   "Effect": "Allow",
+   "Action": [
+    "s3:PutObject",
+    "s3:GetObject",
+    "s3:ListBucket"
+   ],
+   "Resource": [
+    "arn:aws:s3:::%s/*"
+   ]
+  }
+ ]
+}`, bucket))
+	err = s.adm.AddCannedPolicy(ctx, policy, policyBytes)
+	if err != nil {
+		c.Fatalf("policy add error: %v", err)
+	}
+	err = s.adm.SetPolicy(ctx, policy, accessKey, false)
+	if err != nil {
+		c.Fatalf("Unable to set policy: %v", err)
+	}
+	// 2.3 check user has access to bucket
+	c.mustListObjects(ctx, uClient, bucket)
+	// 2.3 check that user cannot delete the bucket
+	err = uClient.RemoveBucket(ctx, bucket)
+	if err == nil || err.Error() != "Access Denied." {
+		c.Fatalf("bucket was deleted unexpectedly or got unexpected err: %v", err)
+	}
+
+	// 3. Craft a request to update the user's permissions
+	ep := s.adm.GetEndpointURL()
+	urlValue := url.Values{}
+	urlValue.Add("accessKey", accessKey)
+	u, err := url.Parse(fmt.Sprintf("%s://%s/minio/admin/v3/add-user?%s", ep.Scheme, ep.Host, s3utils.QueryEncode(urlValue)))
+	if err != nil {
+		c.Fatalf("unexpected url parse err: %v", err)
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, u.String(), nil)
+	if err != nil {
+		c.Fatalf("unexpected new request err: %v", err)
+	}
+	reqBodyArg := madmin.UserInfo{
+		SecretKey:  secretKey,
+		PolicyName: "consoleAdmin",
+		Status:     madmin.AccountEnabled,
+	}
+	buf, err := json.Marshal(reqBodyArg)
+	if err != nil {
+		c.Fatalf("unexpected json encode err: %v", err)
+	}
+	buf, err = madmin.EncryptData(secretKey, buf)
+	if err != nil {
+		c.Fatalf("unexpected encryption err: %v", err)
+	}
+
+	req.ContentLength = int64(len(buf))
+	sum := sha256.Sum256(buf)
+	req.Header.Set("X-Amz-Content-Sha256", hex.EncodeToString(sum[:]))
+	req.Body = ioutil.NopCloser(bytes.NewReader(buf))
+	req = signer.SignV4(*req, accessKey, secretKey, "", "")
+
+	// 3.1 Execute the request.
+	resp, err := s.TestSuiteCommon.client.Do(req)
+	if err != nil {
+		c.Fatalf("unexpected request err: %v", err)
+	}
+	if resp.StatusCode != 200 {
+		c.Fatalf("got unexpected response: %#v\n", resp)
+	}
+
+	// 3.2 check that user cannot delete the bucket
+	err = uClient.RemoveBucket(ctx, bucket)
+	if err == nil || err.Error() != "Access Denied." {
+		c.Fatalf("User was able to escalate privileges (Err=%v)!", err)
 	}
 }
 

--- a/cmd/iam-store.go
+++ b/cmd/iam-store.go
@@ -1629,7 +1629,7 @@ func (store *IAMStoreSys) ListServiceAccounts(ctx context.Context, accessKey str
 }
 
 // AddUser - adds/updates long term user account to storage.
-func (store *IAMStoreSys) AddUser(ctx context.Context, accessKey string, uinfo madmin.UserInfo) error {
+func (store *IAMStoreSys) AddUser(ctx context.Context, accessKey string, ureq madmin.AddOrUpdateUserReq) error {
 	cache := store.lock()
 	defer store.unlock()
 
@@ -1642,9 +1642,9 @@ func (store *IAMStoreSys) AddUser(ctx context.Context, accessKey string, uinfo m
 
 	u := newUserIdentity(auth.Credentials{
 		AccessKey: accessKey,
-		SecretKey: uinfo.SecretKey,
+		SecretKey: ureq.SecretKey,
 		Status: func() string {
-			if uinfo.Status == madmin.AccountEnabled {
+			if ureq.Status == madmin.AccountEnabled {
 				return auth.AccountOn
 			}
 			return auth.AccountOff
@@ -1657,25 +1657,7 @@ func (store *IAMStoreSys) AddUser(ctx context.Context, accessKey string, uinfo m
 
 	cache.iamUsersMap[accessKey] = u.Credentials
 
-	// Set policy if specified.
-	if uinfo.PolicyName != "" {
-		policy := uinfo.PolicyName
-		// Handle policy mapping set/update
-		mp := newMappedPolicy(policy)
-		for _, p := range mp.toSlice() {
-			if _, found := cache.iamPolicyDocsMap[policy]; !found {
-				logger.LogIf(GlobalContext, fmt.Errorf("%w: (%s)", errNoSuchPolicy, p))
-				return errNoSuchPolicy
-			}
-		}
-
-		if err := store.saveMappedPolicy(ctx, accessKey, regUser, false, mp); err != nil {
-			return err
-		}
-		cache.iamUserPolicyMap[accessKey] = mp
-	}
 	return nil
-
 }
 
 // UpdateUserSecretKey - sets user secret key to storage.

--- a/cmd/iam.go
+++ b/cmd/iam.go
@@ -954,7 +954,7 @@ func (sys *IAMSys) DeleteServiceAccount(ctx context.Context, accessKey string, n
 
 // CreateUser - create new user credentials and policy, if user already exists
 // they shall be rewritten with new inputs.
-func (sys *IAMSys) CreateUser(ctx context.Context, accessKey string, uinfo madmin.UserInfo) error {
+func (sys *IAMSys) CreateUser(ctx context.Context, accessKey string, ureq madmin.AddOrUpdateUserReq) error {
 	if !sys.Initialized() {
 		return errServerNotInitialized
 	}
@@ -967,11 +967,11 @@ func (sys *IAMSys) CreateUser(ctx context.Context, accessKey string, uinfo madmi
 		return auth.ErrInvalidAccessKeyLength
 	}
 
-	if !auth.IsSecretKeyValid(uinfo.SecretKey) {
+	if !auth.IsSecretKeyValid(ureq.SecretKey) {
 		return auth.ErrInvalidSecretKeyLength
 	}
 
-	err := sys.store.AddUser(ctx, accessKey, uinfo)
+	err := sys.store.AddUser(ctx, accessKey, ureq)
 	if err != nil {
 		return err
 	}

--- a/cmd/signature-v4-utils_test.go
+++ b/cmd/signature-v4-utils_test.go
@@ -76,7 +76,7 @@ func TestCheckValid(t *testing.T) {
 		t.Fatalf("unable create credential, %s", err)
 	}
 
-	globalIAMSys.CreateUser(ctx, ucreds.AccessKey, madmin.UserInfo{
+	globalIAMSys.CreateUser(ctx, ucreds.AccessKey, madmin.AddOrUpdateUserReq{
 		SecretKey: ucreds.SecretKey,
 		Status:    madmin.AccountEnabled,
 	})

--- a/go.mod
+++ b/go.mod
@@ -49,7 +49,7 @@ require (
 	github.com/minio/csvparser v1.0.0
 	github.com/minio/highwayhash v1.0.2
 	github.com/minio/kes v0.14.0
-	github.com/minio/madmin-go v1.1.18
+	github.com/minio/madmin-go v1.1.20
 	github.com/minio/minio-go/v7 v7.0.17
 	github.com/minio/parquet-go v1.1.0
 	github.com/minio/pkg v1.1.9

